llvm/test/Analysis/ScalarEvolution/max-backedge-taken-count-guard-info.ll

```diff
@@ -1595,6 +1595,145 @@ exit:
+; CHECK-NEXT:    %ptr.iv.next = getelementptr inbounds i8, ptr %ptr.iv, i64 8
+; CHECK-NEXT:    --> {(8 + %a),+,8}<nuw><%loop> U: full-set S: full-set Exits: (8 + (8 * ((-8 + (-1 * (ptrtoint ptr %a to i64)) + (ptrtoint ptr %b to i64)) /u 8))<nuw> + %a) LoopDispositions: { %loop: Computable }
+; CHECK-NEXT:  Determining loop execution counts for: @ptr_induction_eq_1
+; CHECK-NEXT:  Loop %loop: backedge-taken count is ((-8 + (-1 * (ptrtoint ptr %a to i64)) + (ptrtoint ptr %b to i64)) /u 8)
+; CHECK-NEXT:  Loop %loop: constant max backedge-taken count is i64 2305843009213693951
+; CHECK-NEXT:  Loop %loop: symbolic max backedge-taken count is ((-8 + (-1 * (ptrtoint ptr %a to i64)) + (ptrtoint ptr %b to i64)) /u 8)
+; CHECK-NEXT:  Loop %loop: Trip multiple is 1
+;
+entry:
+  %cmp = icmp eq ptr %a, %b
+  br i1 %cmp, label %exit, label %loop
+
+loop:
+  %ptr.iv = phi ptr [ %ptr.iv.next, %loop ], [ %a, %entry ]
+  %ptr.iv.next = getelementptr inbounds i8, ptr %ptr.iv, i64 8
+  %exitcond = icmp eq ptr %ptr.iv.next, %b
+  br i1 %exitcond, label %exit, label %loop
+
+exit:
+  ret void
+}
+
+define void @ptr_induction_eq_2(ptr %a, i64 %n) {
+; CHECK-LABEL: 'ptr_induction_eq_2'
+; CHECK-NEXT:  Classifying expressions for: @ptr_induction_eq_2
+; CHECK-NEXT:    %b = getelementptr inbounds ptr, ptr %a, i64 %n
+; CHECK-NEXT:    --> ((8 * %n)<nsw> + %a) U: full-set S: full-set
+; CHECK-NEXT:    %ptr.iv = phi ptr [ %ptr.iv.next, %loop ], [ %a, %entry ]
+; CHECK-NEXT:    --> {%a,+,8}<nuw><%loop> U: full-set S: full-set Exits: ((8 * ((-8 + (8 * %n)<nsw>) /u 8))<nuw> + %a) LoopDispositions: { %loop: Computable }
+; CHECK-NEXT:    %ptr.iv.next = getelementptr inbounds i8, ptr %ptr.iv, i64 8
+; CHECK-NEXT:    --> {(8 + %a),+,8}<nuw><%loop> U: full-set S: full-set Exits: (8 + (8 * ((-8 + (8 * %n)<nsw>) /u 8))<nuw> + %a) LoopDispositions: { %loop: Computable }
+; CHECK-NEXT:  Determining loop execution counts for: @ptr_induction_eq_2
+; CHECK-NEXT:  Loop %loop: backedge-taken count is ((-8 + (8 * %n)<nsw>) /u 8)
+; CHECK-NEXT:  Loop %loop: constant max backedge-taken count is i64 2305843009213693951
+; CHECK-NEXT:  Loop %loop: symbolic max backedge-taken count is ((-8 + (8 * %n)<nsw>) /u 8)
+; CHECK-NEXT:  Loop %loop: Trip multiple is 1
+;
+entry:
+  %b = getelementptr inbounds ptr, ptr %a, i64 %n
+  %cmp = icmp eq ptr %a, %b
+  br i1 %cmp, label %exit, label %loop
+
+loop:
+  %ptr.iv = phi ptr [ %ptr.iv.next, %loop ], [ %a, %entry ]
+  %ptr.iv.next = getelementptr inbounds i8, ptr %ptr.iv, i64 8
+  %exitcond = icmp eq ptr %ptr.iv.next, %b
+  br i1 %exitcond, label %exit, label %loop
+
+exit:
+  ret void
+}
+
+; TODO: It feels like we should be able to calculate the symbolic max
+; exit count for the loop.inc block here, in the same way as
+; ptr_induction_eq_1. The problem seems to be in howFarToZero when the
+; ControlsOnlyExit is set to false.
```

Comment:
For ptr_induction_eq_1, we prove that %b-%a is divisible by 8, so the trip count is (b-a)/8.  (If difference isn't divisible by 8, we eventually branch on poison.)  Here, we can't prove that: as long as there's another exit, the loop can go on indefinitely (at least, until we read past the end of the allocation).  So we can't do any useful math.

If you tell the compiler the pointers are aligned (for example, using an "align" attribute), we do the computation.

Alternatively, we could compute a count assuming the loop exits via a particular branch; see https://github.com/llvm/llvm-project/pull/83052#pullrequestreview-2013190120 .

Reply:
Yep, this looks to be the same underlying issue as for https://github.com/llvm/llvm-project/issues/101372. Looking into why Clang doesn't generate `align` attributes and what would be needed to change that.

Reply:
But in that example, unless I'm missing something, the early exit dominates the latch. I don't see how the early exit could make the symbolic max loop trip count greater than the backedge-taken count from the latch? I'm probably not explaining myself that well, but what I mean is surely the early exit can only reduce the trip count, not extend it?

Reply:
IIUC, consider the something like `%b = N * 8 * %a + 1`, so `%exitcond` is never true, so we never exit the loop through it.

The early exit then could exit at any later iteration (e.g. when ptr iv is `(N + 2) * 8 * %a`) , before `%ptr.iv.next` wraps.

Reply:
That's fair enough, but then surely the loop in `ptr_induction_eq_1` could also go on indefinitely and hence have an infinite symbolic max backedge-taken count? From what you're saying it sounds like there is a bug and ptr_induction_eq_1 should also return COULDNOTCOMPUTE?

Reply:
The getelementptr operation does have inbounds on it - does that help in this case limit the trip count because it cannot wrap?

Reply:
Anyway, I'm probably missing something obvious here and I'll go off and think about it some more. :)

Reply:
`ptr_induction_eq_1` is different as there's no early exit, so we *must* exit via the check of `ptr.iv.next`. `%b = N * 8 * %a + 1` would trigger UB in `ptr_induction_eq_1`, as %ptr.iv.next` is guaranteed to wrap if the single exit isn't taken.

The early exit in `ptr_induction_early_exit_eq_1` allows for scenarios where `ptr.iv > %b` while still not triggering UB because then we must exit via the early exit before wrap.

Reply:
Ah I see. That's the bit I was missing. I *think* I understand this now. Thanks! So if the IR contained information about alignment that might help?

Reply:
Anyway, the reason I'm asking about this is because I believe this to be a common pattern when you invoke the C++ routine `std::find(vec.begin(), vec.end(), needle)` where vec is of type `std::vector`. The `end()` value at least in libc++ is just a pointer, rather than being constructed in terms of `vec.begin() + vec.size()`.

Reply:
> The getelementptr operation does have inbounds on it - does that help in this case limit the trip count because it cannot wrap?

This helps us prove the loop is finite, but we can't produce a symbolic bound; at most, we could prove the max iteration count is 2^60 or something like that.  Which is marginally helpful.

---------

We could also generate a predicated backedge-taken count, which would be useful in contexts like vectorization.  Granted, we don't usually vectorize loops with multiple exits anyway, but I guess that's changing a bit.

Reply:
Had a look at adding the predicate and it looks that also helps in some non-multi-exit cases: https://github.com/llvm/llvm-project/pull/108777


https://github.com/llvm/llvm-project/issues/101372 also contains some discussion about encoding alignment assumptions for certain libc++ data types/interfaces, as that would also help in other cases for which versioning is not applicable (e.g. removing bounds checks)